File: src/models/personal_task.py
from datetime import datetime
from typing import Literal
from pydantic import BaseModel, ConfigDict, Field

CategoryType = Literal["STUDY", "CAREER", "PERSONAL", "HEALTH", "MEETING", "OTHER"]
PriorityType = Literal["LOW", "MEDIUM", "HIGH", "CRITICAL"]
StatusType = Literal["NOT_STARTED", "IN_PROGRESS", "COMPLETED"]


class PersonalTaskCreateRequest(BaseModel):
    title: str = Field(..., min_length=1, max_length=255, description="Task title")
    description: str | None = Field(default=None, description="Task description")
    category: CategoryType = Field(default="STUDY", description="Task category")
    priority: PriorityType = Field(default="MEDIUM", description="Task priority level")
    status: StatusType = Field(default="NOT_STARTED", description="Task progress status")
    estimated_hours: float | None = Field(default=None, ge=0, description="Estimated hours to complete")
    due_date: datetime | str | None = Field(default=None, description="Due date and time")


class PersonalTaskUpdateRequest(BaseModel):
    title: str | None = Field(default=None, min_length=1, max_length=255)
    description: str | None = None
    category: CategoryType | None = None
    priority: PriorityType | None = None
    status: StatusType | None = None
    estimated_hours: float | None = Field(default=None, ge=0)
    due_date: datetime | str | None = None


class PersonalTaskStatusUpdateRequest(BaseModel):
    status: StatusType = Field(..., description="New task progress status")


class PersonalTaskResponse(BaseModel):
    id: str
    student_id: str
    title: str
    description: str | None = None
    category: str = "STUDY"
    priority: str = "MEDIUM"
    status: str = "NOT_STARTED"
    estimated_hours: float | None = None
    due_date: datetime | str | None = None
    created_at: datetime | str
    updated_at: datetime | str

    model_config = ConfigDict(from_attributes=True)

File: frontend/src/types/personalTask.ts
export type TaskCategory = 'STUDY' | 'CAREER' | 'PERSONAL' | 'HEALTH' | 'MEETING' | 'OTHER';
export type TaskPriority = 'LOW' | 'MEDIUM' | 'HIGH' | 'CRITICAL';
export type TaskStatus = 'NOT_STARTED' | 'IN_PROGRESS' | 'COMPLETED';

export interface PersonalTask {
  id: string;
  student_id: string;
  title: string;
  description?: string | null;
  category: TaskCategory;
  priority: TaskPriority;
  status: TaskStatus;
  estimated_hours?: number | null;
  due_date?: string | null;
  created_at: string;
  updated_at: string;
}

export interface PersonalTaskCreatePayload {
  title: string;
  description?: string;
  category?: TaskCategory;
  priority?: TaskPriority;
  status?: TaskStatus;
  estimated_hours?: number;
  due_date?: string;
}

export interface PersonalTaskUpdatePayload {
  title?: string;
  description?: string;
  category?: TaskCategory;
  priority?: TaskPriority;
  status?: TaskStatus;
  estimated_hours?: number;
  due_date?: string;
}

export interface PersonalTaskStatusUpdatePayload {
  status: TaskStatus;
}

export interface PersonalTaskFilterParams {
  status?: string;
  priority?: string;
  category?: string;
  sort_by?: string;
}
